Block API: Add block styles variations API

youknowriad opened the pull request:
This is the first step towards #783 

It adds a `transforms.styles` Block API to support style variations per block. It works by adding an `is-style-{styleName}`  className to the custom className per block. 

This PR replaces the style variations of the quote block by this new API. It keeps the UI similar for now.

**Testing instructions**

 - Add a quote block
 - Switch the quote style

### Visuals

![image](https://user-images.githubusercontent.com/548849/41768153-bc0831da-760b-11e8-808c-f01be68a6f14.png)

youknowriad committed:
Block API: Add block styles variation API

youknowriad committed:
Adding some unit tests

youknowriad committed:
Fix unit tests

paulwilde commented:
It would be nice if rather than having a fixed `is-style-{styleName}` naming convention for each style, if the option existed for assigning a specific class for each style within the `transforms.styles` API.

This would then allow for this API to be utilised for existing websites rather than having to adapt.

youknowriad committed:
Fix e2e tests

youknowriad commented:
> This would then allow for this API to be utilised for existing websites rather than having to adapt.

I wonder if it's a good idea to adapt the API to existing websites. I tend to think while we have the opportunity to change things, we should seek for the best API instead and a naming convention seems better to me for the moment, it makes it clear where the classname comes from.

Also, this doesn't break anything since we have the "deprecated versions" API so block that upgrade to use this new API can adapt pretty easily without breakage.

gziolo commented on core-blocks/quote/index.js:
```diff
@@ -66,6 +60,10 @@ export const settings = {
 	attributes: blockAttributes,
 
 	transforms: {
+		styles: [
```
This is a very expected API 👍

mtias commented on editor/components/block-styles/index.js:
```diff
@@ -0,0 +1,110 @@
+
+	return (
+		<Toolbar controls={ styles.map( ( variation ) => ( {
+			icon: variation.icon,
```
I don't think we'd want to attach the style variation to an icon, since we are likely to drop that approach for quotes. I imagine initially we'd just show a custom button in Quote toolbar to toggle between the registered styles. Once we build the proper UI for style variations we'd just remove the toolbar item.

chrisvanpatten commented:
Two things that are more vague related thoughts than anything else…

1. Some blocks might want a block style variation to change more than just a CSS class. Of course I don't think that's necessarily the API's job to worry about handling all those cases: as long as the block's style value is available outside a `className` context (which it appears to be), it should be easy enough for the block to handle those conditions and transform everything it wants to transform.
2. Right now it looks like the styles have an `icon` parameter. Would that be compatible with @karmatosed's amazing mockup [here](https://github.com/WordPress/gutenberg/issues/783#issuecomment-395740295) where the icon is more like an image?

youknowriad commented:
@mtias @chrisvanpatten yes, the `icon` parameter is here to keep the same UI as today but maybe we should continue the UI work in this PR to get rid of this parameter.

> Some blocks might want a block style variation to change more than just a CSS class. Of course I don't think that's necessarily the API's job to worry about handling all those cases: as long as the block's style value is available outside a className context (which it appears to be), it should be easy enough for the block to handle those conditions and transform everything it wants to transform.

In general, I think if extending block requires more than adding a className or buttons to update existing block attributes, I tend to think we should probably think about creating a separate block transformable to the current block using the transforms API instead of trying to extend the existing blocks.

mtias commented:
> who use this API to provide alternate designs for a block, may need small markup changes between variants within a single block that don't warrant creating an entirely separate block.

The other problem is this introduces too many variations that would be hard to account in validation stages, particularly if the plugin extending a core block is disabled — the block would become invalid, and we wouldn't have knowledge of which transformations to apply.

Restricting variation of variants to classes would allow people to extend blocks without the overhead of creating their own, but without suffering from invalidation hurdles.

chrisvanpatten commented:
@mtias I'm not referring to extending existing blocks, only to custom blocks. I agree completely that if you're extending an existing block, you should be limited to class names. But authors of completely custom blocks might want access (or even _need_ access, in the case of server-rendered blocks) to this attribute.

mtias commented:
Oh, sure, the attribute should totally be exposed as it'd be useful in many scenarios.

youknowriad committed:
Update the block switcher UI to show style variations and previews

youknowriad commented:
Ok, so I've pushed this further and implemented the UI proposed by @karma

Instead of using an "example" to preview the style, I'm using the current block's content.

youknowriad committed:
Always open the block transforms panel and avoid leaking the text-align

mtias commented:
> Instead of using an "example" to preview the style, I'm using the current block's content.

This is clever, good call.

mtias commented:
There is one more thing here that I'd like to ponder, which is how it could relate to the "additional CSS class" panel.

![image](https://user-images.githubusercontent.com/548849/41717889-7bb9d7c4-755b-11e8-86be-b28c8f9246b3.png)

I'm thinking it could be neat to let user write a class and click a button to register that class as a block style variation:

![image](https://user-images.githubusercontent.com/548849/41718061-1a63de7e-755c-11e8-9680-c6b180730b2a.png)

Thinking further, we could even let you write CSS targeting this specific class, and we'd apply it to the block / miniature automatically.

All this is definitely not for this issue :)

youknowriad committed:
Use the same inserterg grid view for block transforms

youknowriad committed:
Fix unit tests

youknowriad committed:
Add style variations e2e test

youknowriad commented on test/e2e/specs/style-variation.test.js:
```diff
@@ -0,0 +1,31 @@
+
+		// Check the content
+		const content = await getHTMLFromCodeEditor();
+		expect( content ).toMatchSnapshot();
```
You mean a regex to catch the className or something? 

I mean we can add it but we do have unit tests to test the functions applying and retrieving those classNames.

gziolo replied:
Probably XPath expression offered by Puppetter API would be sufficient, but yes, this is what I thought about.

gziolo replied:
It's a blurry territory between e2e and unit tests, if it is covered already then it's fine to skip this suggestion.

mtias committed:
Use block icon with colors in main block toolbar.

Move styles to the relevant scopes.

mtias commented:
Pushed change to use the small block icon with colors and drop the down arrow.

![image](https://user-images.githubusercontent.com/548849/41767430-dd10d71c-7609-11e8-9c24-6b0e8bb8a91f.png)

mtias committed:
Adjust padding of block switcher chip to avoid toolbar clipping.

mtias committed:
Center preview for block styles.

mtias commented:
When the popover switches to the top, some of the containing elements break out of it:

![image](https://user-images.githubusercontent.com/548849/41768310-21ec4f86-760c-11e8-853b-ad479d5e8460.png)

mtias commented:
Closes #7227.

youknowriad committed:
Use button to ensure valid HTML

youknowriad commented:
@mtias Did you fix the extra border showing up when the popover switches to the top? I can't reproduce in my testing.

youknowriad committed:
Fix lint

mtias commented:
I didn't.

karmatosed commented:
This looks really great so far thanks for all the work on it.

One thing I did wonder about is there anything we can do about the formatting on the style variants?

![2018-06-22 at 21 27](https://user-images.githubusercontent.com/253067/41798046-1e0234d2-7664-11e8-83e8-edee43fc8758.png)

Above you can see what happens with the default text and it looks weird to have one so small and the other on two lines.

youknowriad commented:
@karmatosed What we can probably do is to give a minimum width to the preview and use `overflow: hidden` to crop it when the available width is smaller than this min width. This would avoid returning to the line too often. 

Though, I think no matter what we do, we'd still have styling issues because the blocks are not meant to be shown in such a small area.

mtias committed:
Add alternate styles for separator.

chrisvanpatten commented:
What's the a11y impact of having the live nested block previews, vs an SVG or raster image? Do screen readers try to read out or navigate the blocks?

youknowriad committed:
Show style variations when the block type doesn't have any block transform

youknowriad committed:
Improve a11y of the style switcher

youknowriad commented:
@chrisvanpatten I've labeled the buttons and applied `aria-hidden` around the preview which would make sure screen readers ignore it IIRC.

mtias committed:
Add Button block style variants and fix line height.

jasmussen committed:
Polish the visuals a bit.

- Stack borders
- Vertically center the contents in the previews
- Tweak hover/focus styles

jasmussen commented:
I pushed a little polish to the visuals:

![screen shot 2018-06-26 at 12 44 48](https://user-images.githubusercontent.com/1204802/41906730-e94a7230-793e-11e8-83b9-8bb74faac42d.png)

The hover/focus style of the chip and previews are polished a bit, and there were some stacking borders I removed.

Noting that there is still a box shadow applied to two elements, both the panel content itself, and the hover preview — this means stacked shadows. The way this is "fixed" in the topmost block library is simply that the scrollbar and search field covers the shadow, it's still there. Not necessarily something we need to fix here, but would be nice if it was _one piece of paper_ instead of _two stacked on top of each other_.

Also, it would be nice with an "applied" style for the block variation you've chosen. I can visually design this style, but the class is missing it seems.

youknowriad commented on editor/components/block-styles/style.scss:
```diff
@@ -2,27 +2,40 @@
 	}
 
+	// Make vertical centering easier.
+	> blockquote.wp-block-quote,
```
I think this is too specific and probably not something we want to fix in the `block-styles` stylesheet.

jasmussen replied:
Open to suggestions, but the issue this fixes is that it enables true vertical centering of the contents inside the preview, by making sure the margins don't offset the balance.

mtias replied:
I think we need to figure out another way for this. I got it almost working with `align-self: center`.

jasmussen replied:
Well, if we added yet another wrapping container element, then we could apply some large margins above and below that element. Through the magic of collapsing margins, the largest margin would win, and because it would be the same above and below, then centering would work.

youknowriad replied:
Disabling the keyboard requires a wrapper, I'm adding a wrapper for the preview not sure if this would fix the issue though.

youknowriad commented on editor/components/block-styles/style.scss:
```diff
@@ -2,27 +2,40 @@
 
 	&:hover {
-		@include formatting-button-style__hover;
+		@include block-style__hover;
```
I personally don't like the block shadow on hover, It feels too dark and not very polished but that's not specific to this PR. (It's similar in the block inserter).

jasmussen replied:
I'd be fine with exploring a different hover style for blocks in general, we can make the shadows lighter or do something different. But let's do that in the mixin stylesheet, so it's consistent for all block buttons.

youknowriad commented on editor/components/block-styles/style.scss:
```diff
@@ -0,0 +1,41 @@
+    display: flex;
+    justify-content: center;
+    align-items: center;
+    flex-direction: column;
```
Mixed tabs/spaces

jasmussen replied:
Fixed, thanks.

jasmussen committed:
Fix tabs/spaces.

youknowriad committed:
Fix keyboard navigation into the styles preview buttons

jasmussen committed:
Polish preview container, fix focus, separator

jasmussen commented:
Pushed a little further polish to fix an issue with the preview after the field became disabled, polishing the separator and getting the preview to work there, and adding a focus style. This focus style doubles as a style for `is-active`, which so far is the only piece missing: an indicator of which variation you're seeing now. Screenshots:

![screen shot 2018-06-26 at 15 07 48](https://user-images.githubusercontent.com/1204802/41914344-124a8a26-7953-11e8-868a-22856edd3473.png)

![screen shot 2018-06-26 at 15 08 23](https://user-images.githubusercontent.com/1204802/41914348-137b61d6-7953-11e8-9537-9c5312ed4dbf.png)

![screen shot 2018-06-26 at 15 08 28](https://user-images.githubusercontent.com/1204802/41914349-14cb682e-7953-11e8-81e1-0d3a3b361287.png)

youknowriad committed:
Adding the is-active className to the block-styles component

youknowriad committed:
Move the styles BlockAPI to the root level (out of the transforms)

jasmussen committed:
Rename CSS class to utilize is-active.

Sorry about the typo.

chrisvanpatten commented:
How would it look if the labels were visible below each variation, like the transforms? I feel like that might help the usability a bit, to be able to quickly find a variation for a specific use-case. Another option could be displaying the label in the Preview header area.

mtias commented:
Either a label below (like the original concept) or using a Tooltip, it's just not added yet.

youknowriad committed:
Showing style variation labels

youknowriad commented:
I've added the labels under the previews, Let me know what you think and what remains to get this in?

Commit:
Polish the separator styles

- Bug: the body text was not the same gray as the title. This fixes that.
- The new Dots separator style is simpler, smaller, better centered, and uses the same color as the text.
- The wide separator style is now thinner.

jasmussen commented:
I pushed a little polish to the separator styles:

<img width="726" alt="screen shot 2018-06-27 at 11 08 24" src="https://user-images.githubusercontent.com/1204802/41964511-753bc896-79fa-11e8-8bbc-e9df8785dda3.png">

Riad, I think the labels look great, however there's something with the CSS grid that means the variation buttons do not have the same width:

<img width="730" alt="screen shot 2018-06-27 at 11 09 50" src="https://user-images.githubusercontent.com/1204802/41964577-a05abdde-79fa-11e8-849b-baf67a83ddf5.png">

I didn't touch it as I'm not super skilled in CSS grid — do you know how to fix this?

youknowriad committed:
Fix grid items width

youknowriad commented:
@jasmussen The items width should be fixed (at least it is for me), let me know if it works for you.

jasmussen commented:
Beeyoutiful:

<img width="460" alt="screen shot 2018-06-27 at 11 21 41" src="https://user-images.githubusercontent.com/1204802/41965183-489460f8-79fc-11e8-9a9e-a2175883da90.png">

Thank you.

mtias commented:
@jasmussen we need to get these styles on the front-end as well for the separator.

mtias committed:
Use proper label names for quote styles.

Commit:
Polish separator styles for frontend

- Rename remaining instance of "asterisks" to "dots"
- Move default separator style to theme.scss so it can be opted into
- Scope default style so it doesn't bleed into dots and wide styles
- Refine dots and wide styles to better work on frontend

jasmussen commented:
Pushed some frontend style polish in https://github.com/WordPress/gutenberg/pull/7362/commits/e34735d5d37e34b8e1b3d54ebf15631d03587ae5

Editor:

<img width="818" alt="screen shot 2018-06-27 at 12 44 19" src="https://user-images.githubusercontent.com/1204802/41969530-01609ac4-7a08-11e8-8178-40a2d566f106.png">

Frontend:

<img width="995" alt="screen shot 2018-06-27 at 12 44 14" src="https://user-images.githubusercontent.com/1204802/41969538-0600660e-7a08-11e8-9779-6b7433688554.png">

Note, the wide style here is limited due to a bug in the gutenberg starter theme where we simply have to remove it, I've verified that it's not an issue with the theme CSS.

Please have a look at the CSS pushed to see that it's kosher.

mtias committed:
Add button outline style.

youknowriad committed:
Fix unit tests (Block Switcher without blocks)

youknowriad commented:
I rebased. I'm seeing intermittent e2e test failures preventing the merge. Let's see how it goes this time.

aduth commented on editor/components/block-styles/index.js:
```diff
@@ -0,0 +1,155 @@
+		<div className="editor-block-styles">
+			{ styles.map( ( style ) => {
+				const styleClassName = replaceActiveStyle( className, activeStyle, style );
+				/* eslint-disable jsx-a11y/click-events-have-key-events */
```
This is an unfortunate instance of blindly disabling a rule for which its entire purpose is to protect against the thing we've introduced here.

https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/click-events-have-key-events.md
